File: Python/프로그래머스/level_2/피로도.py
# 일정 피로도를 사용해서 던전을 탐험
# 최소 필요 피로도, 소모 피로도

# 최대한 많은 던전을 탐험

# 던전 갯수는 1 이상 8 이하
# 각 행은 [최소 필요 피로도, 소모 피로도]

# 탐험할 수 있는 최대 던전 수는?

# 남은 던전들 중에서 최소 필요 피로도를 만족하면서 소모 피로도가 가장 작은걸 매번 골라주면 되겠다
# 아님 왜냐하면 어떤 던전 하나를 가서 다른 던전의 최소필요피로도를 만족 못하는 경우가 발생
# 그러면 걍 permutations으로 다 돌아야하나?

from itertools import permutations


def solution(k, dungeons):
    answer = -1
        
    
    
    num_list = [i for i in range(len(dungeons))]
    perms = list(permutations(num_list))
    
    
    max_val = 0
    for perm in perms:
        pyro = k
        cnt = 0
        for i in range(len(perm)):
            least_pyro, consume_pyro = dungeons[perm[i]]
            if least_pyro <= pyro:
                pyro -= consume_pyro
                cnt += 1
        max_val = max(max_val, cnt)
    
    answer = max_val
    return answer



print(solution(80, [[80,20],[50,40],[30,10]])) # 3

File: Kotlin/프로그래머스/Level 2/피로도.kt
// -던-
// 탐험을 시작하기 위한 "최소 필요 피로도", 던전 탐험을 마쳤을 떄 소모되는 "소모 피로도"
// 최대한 많은 던전을 탐험
// 근데 dungeons 의 갯수가 8개 밖에 안되서 다 permutation 돌려도 될 듯?

class Solution {
    var maxVal = 0 // 가장 많이 돌은 던전 기록
    
    fun solution(k: Int, dungeons: Array<IntArray>): Int {
        var answer: Int = -1
        
        
        
        var visited = BooleanArray(dungeons.size)
        permutation(dungeons, visited, k, 0)
        
        answer = maxVal
        
        return answer
    }
    
    fun permutation(dungeons:Array<IntArray>, visited:BooleanArray, k:Int, cnt:Int):Unit {
        if(visited.count{it == true} == visited.size){ // 재귀 탈출조건
            if(cnt > maxVal){
                maxVal = cnt
            }
            return
        }
        
        for(i in 0 until visited.size){
            if(visited[i] == false){ // 방문 안한 던전인데
                
                visited[i] = true // 방문 전처리
                
                if(dungeons[i][0] <= k){ // 방문 가능
                    permutation(dungeons, visited, k-dungeons[i][1], cnt+1)
                }
                else{
                    permutation(dungeons, visited, k, cnt)
                }
                
                visited[i] = false // 방문 후처리
                
            }
        }
    }
}